####Problem: 1
Find the factorial of given "n", using recursion.

In [ ]:
def factorial(n):

  #edge case when n = 0:
  if n == 0:
    return 1
  else:
    return factorial(n-1) * n

In [ ]:
if __name__ == '__main__':
  n = 7
  value = factorial(n)
  print(n,"!:",value)

7 !: 5040


### Recursion Trace

factorial(5) $: 24 * 5 = 120$

> factorial(4) $: 6 * 4 = 24$

>> factorial(3)$: 2 * 3 = 6 $

>>> factorial(2) $: 1 * 2 = 2$

>>>> factorial(1) $: 1*1 = 1$

>>>>> factorial(0): $1$


####Problem: 2
Implement binary search, to find the data in a list, using recursion.

Notes: 
Binary search, is used to effectively locate a target value within a sorted sequence of n elements.This is among the most important computer algorithms, and it is  the reason that we so often store data in sorted order.

When the sequence is **unsorted**, the standard approach to search for a target  value is to loop to examine every element, until either finding the target or exhausting the data set.  This is known as **Sequential Search Algorithm**. This algorithm runs **$ Ο(n)$**

In [ ]:
def binary_search(target,arr,low,high):
  """
  Return True if target is found in indicated portion of Python list. 
  The search only considers the portion from data[low] to data[high]
  """

  if low > high:
    return False   # Interval is empty no match
  else:
    mid = (low + high)// 2
    if target == arr[mid]:
      return True
    elif target > arr[mid] :
      # recur on the portion right of the middle
      return binary_search(target,arr,mid+1,high)
    else:
      # recur on the portion left of the middle
      return binary_search(target,arr,low,mid-1)

In [ ]:
arr = [2,4,5,7,8,12,14,17,19,22,25,27,28,33,37]

target= 22
low = 0
high = len(arr)-1
result = binary_search(target,arr,low,high)

if result is True:
  print("The given target data is present in the given list" )
else:
  print("The given number target data is not present in the list")

The given target data is present in the given list


### Recursion Trace 

Subtituting with zero's (just for understanding, how the search region cuts in half, for every recursion call

binary_search\
$1^{st}$Function call $: [2,4,5,7,8,12,14,17,19,22,25,27,28,33,37]$\
$2^{nd}$Recursive function call$:  [0,0,0,0,0,0,0,0,0,19,22,25,27,28,33,37]$\
$3^{rd}$Recursive Function call $: [0,0,0,0,0,0,0,0,0,19,22,25,0,0,0,0]$\
$4^{th}$Recursive Function call $: [0,0,0,0,0,0,0,0,0,0,22,0,0,0,0,0]$

####Problem: 3
**English Ruler** : Page 175  solved example of the book

$----0$\
$-$\
$--$  
$-$\
$---$\
$-$\
$--$  
$-$\
$----1$\
$-$\
$--$  
$-$\
$---$\
$-$\
$--$  
$-$\
$----2$


A Recursive approach to Ruler drawing 

The english ruler pattern is a simple example  of a fractal, that is, a shape that has a self-recursive at various stages of maginification.
Ignoring the lines containing 0,1,2. let us consider how to draw  the sequence of ticks lying between the lines. The central tick between 0 and 1, had length 3. Observe that the two patterns of ticks above and below this tick are identical and has a central tick of length 2.

In general, an interval with a central tick of length $L \geq 1$ is composed of :      
- An interval with central tick length $L-1$
- A single tick of length $L$.
- An interval with a central tick of length $L-1$

Our implementation consists of three functions:     

1. Main Function :  `draw_ruler` : Manages the construction of the entire ruler. Its arguments specify the the total number of inches/cms in the ruler and the major tick.

2. Utility Function : `draw_line`: Draws a single tick with a specified number of dahses.

3. Recursive function:`draw_interval` : This function draws the sequence of minor ticks within some interval,based upon the length of the interval's central tick.

In [ ]:
def draw_line(tick_length,tick_label=''):
  """
  Draw one line with given tick length( followed by optional label)
  """
  line = '-' * tick_length
  if tick_label:
    line+=' ' + tick_label
  print(line)

def draw_interval(center_length):
  """
  Draw tick interval based upon a central tick length
  """
  if center_length >0:                       #stops when length drops to zero
    draw_interval(center_length - 1)          #recursively draw top ticks
    draw_line(center_length)                 #Draw central tick
    draw_interval(center_length - 1)         # recursively draws bottom ticks.

def draw_ruler(num_inches,major_length):
  """
  Draw english ruler with given number of inches, major tick length
  """
  draw_line(major_length,'0')               #draw inch 0 line
  for j in range(1,num_inches+1):
    draw_interval(major_length-1)           #draw interior ticks for inch
    draw_line(major_length, str(j))         #draw inch j line and label


In [ ]:
draw_ruler(2,4)

---- 0
-
--
-
---
-
--
-
---- 1
-
--
-
---
-
--
-
---- 2


### Problem : 4

Reversing a list using recursion

In [ ]:
def reverse(arr,start,stop):
  if start < stop:
    arr[start],arr[stop] = arr[stop], arr[start]
    reverse(arr,start+1,stop-1)

In [ ]:
arr = [4,3,6,2,4,8,9,5]
start = 0 
stop = len(arr)-1
reverse(arr,start,stop)
print("Reversed arr",arr)

Reversed arr [5, 9, 8, 4, 2, 6, 3, 4]


### Problem : 5

Summing the elements of a sequence recursively

In [ ]:
def linear_sum(arr):
  n = len(arr)
  if n == 0:
    return 0
  else:
   return linear_sum(arr[0:n-2]) + arr[n-1]

In [ ]:
arr = [4,3,6,2,8,9,3,2,8,5,1,7,2,8,3,7]
sum = linear_sum(arr)
print("The sum of the elements of the sequence", sum)

The sum of the elements of the sequence 43


### Recursion Trace 
For first 5 elements of the arr

[4,3,6,2,8,9]\
$\text{linear_sum}(arr[0:4]) + arr[5]$     : $23 + 9$

>$\text{linear_sum}(arr[0:3]) + arr[4]$     : $15 + 8$

>> $\text{linear_sum}(arr[0:2]) + arr[3]$  : $13 + 2$

>> $\text{linear_sum}(arr[0:1]) + arr[2]$ : $7 + 6$

>>> $\text{linear_sum}(arr[0]) + arr[1]$   : $4 + 3$

>>> $\text{linear_sum}(arr[])+arr[0]$      : $0 + 4$





### Problem : 6
 Recursive Algorithms for computing power

We consider the problem of raising a a number $x$ to an arbitrary integer $n$. That is we wish to compute the **power function**, *defined* as **$pow(x,n)=x^n$**. We will use two different recursive formulas for the problem.

1. Naive implementaton of power function
2. Advanced implementation of the power function.

In [ ]:
#Naive Implementation, similar to Factorial : Our first recursive algorithm.
def pow_naive(x,n):

  if n == 0:
    return 1
  else:
    return pow(x,n-1) * x

In [ ]:
def pow(x,n):

  if n == 0:
    return 1
  else:
    partial = pow(x,n//2)
    print("partial",partial)
    result = partial * partial 
    if n%2 == 1:
      result *= x
    return result

In [ ]:
x = 2
n = 13
result = pow(x,n)
print(result)

partial 1
partial 2
partial 8
partial 64
8192


###Problem : 7 

**Binary Recursion** :  When a function makes two recursive calls, we say that it uses binary recursion. Computing the sum of zero or one elements is trivial. With two or more elements, we can recursively compute  the sum of the first half, and the sum of the second half  and add the two sums together. 

In [ ]:
def binary_summation(arr,low,high):

  if low > high:       # 0 elements in the list
    return 0
  elif low == high:
    return arr[low]
  else:
    mid = (low+high)//2
    bottom = binary_summation(arr,low,mid)
    top = binary_summation(arr,mid+1,high)
    return bottom + top 

In [ ]:
arr = [6,44,33,22,11,66,43,21]
low = 0
high = len(arr) - 1
binary_summation(arr,low,high)

246

### Problem : 8

Describe a recursive algorithm for finding the maximum element in a sequence, $S$, of $n$ elements. What is your running time and space usage? 

In [ ]:
def bin_max(arr,start,stop):

  if start > stop:        # the array/list is empty
    return 0
  elif start == stop:
    return arr[start]
  else:
    mid = (start + stop) //2
    first_half = bin_max(arr,start,mid)
    second_half = bin_max(arr,mid+1,stop)
    if first_half >= second_half:
      return first_half
    else:
      return second_half

In [ ]:
arr = [1,2,1,0,50,9,8,30,40,0,3,2,4,1,6,80]
start = 0
stop = len(arr) - 1
result = bin_max(arr,start,stop)
print("The maximum element in the sequence S of n elements :",result)

The maximum element in the sequence S of n elements : 80


### Problem : 9

Draw the recursion trace for the computation of **$pow(2,18)$**, using the repeated squaring algorithm

In [ ]:
def pow(x,n):

  if n == 0:
    return 1
  else:
    partial = pow(x,n//2)
    result = partial * partial 
    if n % 2 == 1:
      result = partial * partial * x
    return result 


In [ ]:
x = 2
n = 18
pow(x,n)

262144

$(2,18): 512 * 512   = 262144$
>$(2,9) : 16 * 16 * 2   = 512$
>>$(2,4) : 4 * 4 = 16$
>>>$(2,2) : 2*2 = 4$
>>>>$(2,1) :1*1*2 $
>>>>>$(2,0) : 1$



### Problem : 10

Draw the recursion trace for the execution of the function reverse $(S,0,5)$ on $S=[4,3,6,2,6]$

In [ ]:
def reverse(S,start,stop):

  if start > stop:      #the list is empty 
    return False 
  else:
    S[start],S[stop] = S[stop], S[start]
    return reverse(S,start+1,stop-1)

In [ ]:
S = [9,4,3,0,1,3]
stop = len(S) - 1
reverse(S,0,stop)
print("The reversed list S",S)

The reversed list S [3, 1, 0, 3, 4, 9]


### Problem : 11
Write a short recursive Python function  that finds the minimum and maximum values  in a sequence without using any loops.

In [ ]:
def min_max(S,start,stop):

  if start > stop:                # the given list is empty 
    return 0
  elif start == stop-1:           # When we have only two elements left in the sub-array
    if S[start] > S[stop]:        # Make a comparison among the two, and designate min and max among them 
      min = S[stop]
      max = S[start]
    else:
      min = S[start]
      max = S[stop]
    return min,max
  else:
    mid = (start + stop)//2        # calculat the mid 
    min1,max1 = min_max(S,start,mid)
    min2,max2 = min_max(S,mid,stop)

    if min1 < min2:               # when both the recursion calls are done, compare the min against min, and max against max
      min = min1                  # and return the min and max as per the conditions satisfied.
    else:
      min = min2
    
    if max1 > max2:
      max = max1
    else:
      max = max2
    return min,max

In [ ]:
S = [11,2,0,3,4,4,4,3,3,55,66,2,2,4,3,2,1,0,100,55,43,54,32,101,43,32,904,54]
start = 0
stop =len(S) - 1
min,max = min_max(S,start,stop)
print("The minimum value in the sequence :",min)
print("The maximum value in the sequence :",max)

The minimum value in the sequence : 0
The maximum value in the sequence : 904


### Problem : 12

Give a recursive algorithm to compute the product of two positive integers ,$m$ and $n$, using only adddition and subtraction 

In [ ]:
def prod(m,n):

  if n == 0:
    return 0
  else:
    if m > n:                 # if m > n: means logically, we repeat (n times): lesser number of recursion call
      return m + prod(m,n-1)
    else:                      #else: interchange the values of m and n, to facilitate, lower number of recursion calls.
      return n + prod(n,m-1)

In [ ]:
m = 100
n = 10
multi = prod(m,n)
print("The value of m*n :",multi)

The value of m*n : 1000



Give an input string $s$ and a pattern $p$, implement regular expression matching with suppoer for $"."$ and $"*"$ where:    

$"."$ Matches any single string

$"*"$ Matches zero of more of the preceding element 


In [ ]:
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
         self.val = val
         self.next = next

In [ ]:
head = ListNode(1,None)
head.next = ListNode(2,None)
head.next.next = ListNode(3,None)
head.next.next.next = ListNode(4,None)

In [ ]:
def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
    
    #edge case when both list are empty 
    
    if list1 is None and list2 is None:
        return None
    
      #edge case when one of the list is empty 
        
    if list1 is None:
        return list2
    else:
        return list1
        
    mergelist = ListNode()
    #for the remaining cases where both lists have 1 or more elements
    
    while list1.val is not None or list2.val is not None:
        
        # handle cases where both lists are not empty 
        print("list1",list1.val)
        print("list2",list2.val)
        if list.val is not None and list2.val is not None:
            
            if list1.val == list2.val:
                
                mergelist.val = list1.val
                mergelist = mergelist.next
                mergelist.val = list2.val
                list1 = list1.next
                list2 = list2.next 
                
            elif list1.val > list2.val:
                mergelist.val = list2.val
                list2 = list2.next
            else:   
                mergelist.val = list1.val
                list2 = list2.next        
            
        # handle cases when one of the list is null
        elif list1.val  is None:
            mergelist.val = list2.val
            list2 = list2.next 
            
        else:
            mergelistval = list1.val
            list1 = list1.next
        
        mergelist = merge_list.next    
        
        
        
            

In [ ]:
def swapPairs(head: ListNode):
    
    #Edge case
    if head is None:           # When linklist is empty .
      return None
    
    # Swapping the firt two nodes
    
    linkedlist = head.next 
    tail = head
    head = head.next 
    tail.next = head.next 
    head.next = tail
    
    # swapping for rest of the nodes in the linked list.
    
    while head is not None:
        
        head = tail.next 
        if head.next is not None:
            tail.next = head.next 
            tail = tail.next    
        else:
            tail.next = head 
            return linkedlist
    
    return linkedlist

In [ ]:
swapPairs(head)

KeyboardInterrupt: ignored